android/src/main/java/com/marigold/rnsdk/RNMarigoldModule.kt

```diff
@@ -0,0 +1,330 @@
+                        array.pushMap(jsonConverter.convertJsonToMap(messageJson))
+                    }
+                    promise.resolve(array)
+                } catch (e: NoSuchMethodException) {
+                    promise.reject(ERROR_CODE_MESSAGES, e.message)
+                } catch (e: IllegalAccessException) {
+                    promise.reject(ERROR_CODE_MESSAGES, e.message)
+                } catch (e: JSONException) {
+                    promise.reject(ERROR_CODE_MESSAGES, e.message)
+                } catch (e: InvocationTargetException) {
+                    promise.reject(ERROR_CODE_MESSAGES, e.message)
+                }
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(ERROR_CODE_MESSAGES, error.message)
+            }
+        })
+    }
+
+    // Moved out to separate method for testing as WritableNativeArray cannot be mocked
+    fun getWritableArray(): WritableArray {
+        return WritableNativeArray()
+    }
+
+    @ReactMethod
+    fun getUnreadCount(promise: Promise) {
+        messageStream.getUnreadMessageCount(object : MessageStream.MessageStreamHandler<Int> {
+            override fun onSuccess(value: Int) {
+                promise.resolve(value)
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(ERROR_CODE_MESSAGES, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun removeMessage(messageMap: ReadableMap, promise: Promise) {
+        val message = try {
+            getMessage(messageMap)
+        } catch (e: JSONException) {
+            promise.reject(ERROR_CODE_MESSAGES, e.message)
+            return
+        } catch (e: NoSuchMethodException) {
+            promise.reject(ERROR_CODE_MESSAGES, e.message)
+            return
+        } catch (e: IllegalAccessException) {
+            promise.reject(ERROR_CODE_MESSAGES, e.message)
+            return
+        } catch (e: InvocationTargetException) {
+            promise.reject(ERROR_CODE_MESSAGES, e.message)
+            return
+        } catch (e: InstantiationException) {
+            promise.reject(ERROR_CODE_MESSAGES, e.message)
+            return
+        }
+        messageStream.deleteMessage(message, object : MessageStream.MessageDeletedHandler {
+            override fun onSuccess() {
+                promise.resolve(null)
+            }
+
+            override fun onFailure(error: Error) {
+                promise.reject(ERROR_CODE_MESSAGES, error.message)
+            }
+        })
+    }
+
+    @ReactMethod
+    fun registerMessageImpression(typeCode: Int, messageMap: ReadableMap) {
+        val type = when (typeCode) {
+            0 -> ImpressionType.IMPRESSION_TYPE_IN_APP_VIEW
+            1 -> ImpressionType.IMPRESSION_TYPE_STREAM_VIEW
+            2 -> ImpressionType.IMPRESSION_TYPE_DETAIL_VIEW
+            else -> return
+        }
+        val message = try {
```

Comment:
Same problem as the `Purchase` exception handling above and I would recommend a similar change.